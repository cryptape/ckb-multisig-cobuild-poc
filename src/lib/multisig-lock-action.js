import { ckbHasher } from "@ckb-cobuild/ckb-hasher";
import { Script } from "@ckb-cobuild/ckb-molecule-codecs";
import { ScriptInfo } from "@ckb-cobuild/cobuild";
import mol, { createUint8ArrayJsonCodec, toJson } from "@ckb-cobuild/molecule";
import { SECP256K1_MULTISIG_CODE_HASH } from "./ckb-address.js";

export const PubkeyHash = createUint8ArrayJsonCodec(
  mol.byteArray("PubkeyHash", 20),
);
export const PubkeyHashVec = mol.vector("PubkeyHashVec", PubkeyHash);

export const Signature = createUint8ArrayJsonCodec(
  mol.byteArray("Signature", 65),
);
export const PubkeyHashSignaturePair = mol.struct(
  "PubkeyHashSignaturePair",
  {
    pubkey_hash: PubkeyHash,
    signature: Signature,
  },
  ["pubkey_hash", "signature"],
);
export const PubkeyHashSignaturePairVec = mol.vector(
  "PubkeyHashSignaturePairVec",
  PubkeyHashSignaturePair,
);

export const MultisigConfig = mol.table(
  "MultisigConfig",
  {
    require_first_n: mol.byte,
    threshold: mol.byte,
    signer_pubkey_hashes: PubkeyHashVec,
  },
  ["require_first_n", "threshold", "signer_pubkey_hashes"],
);

export const MultisigAction = mol.table(
  "MultisigAction",
  {
    config: MultisigConfig,
    signed: PubkeyHashSignaturePairVec,
  },
  ["config", "signed"],
);

export function buildScriptInfo() {
  return {
    name: "secp256k1_blake160_multisig_all",
    url: "https://github.com/nervosnetwork/ckb-system-scripts/blob/master/c/secp256k1_blake160_multisig_all.c",
    // Always use empty args to generate script hash
    script_hash: ckbHasher()
      .update(
        Script.pack({
          code_hash: SECP256K1_MULTISIG_CODE_HASH,
          hash_type: "type",
          args: Uint8Array.of(),
        }),
      )
      .digest(),
    schema: Array.from(MultisigAction.exportSchema().values()).join("\n"),
    message_type: "MultisigAction",
  };
}

export const SCRIPT_INFO = buildScriptInfo();
export const SCRIPT_INFO_HASH = ckbHasher()
  .update(ScriptInfo.pack(SCRIPT_INFO))
  .digest();

export function buildAction(lockArgs, actionData) {
  return {
    script_info_hash: SCRIPT_INFO_HASH,
    script_hash: ckbHasher()
      .update(
        Script.pack({
          code_hash: SECP256K1_MULTISIG_CODE_HASH,
          hash_type: "type",
          args: lockArgs,
        }),
      )
      .digest(),
    data: MultisigAction.pack(actionData),
  };
}
export function multisigStatus(actionData) {
  if (actionData.signed.length >= actionData.config.threshold) {
    for (const pubkeyHash of actionData.config.signer_pubkey_hashes.slice(
      0,
      actionData.config.require_first_n,
    )) {
      const pubkeyHashHex = toJson(pubkeyHash);
      if (
        actionData.signed.findIndex(
          (item) => toJson(item.pubkey_hash) === pubkeyHashHex,
        ) === -1
      ) {
        return "partially signed";
      }
    }
    return "ready";
  }

  return actionData.signed.length > 0 ? "partially signed" : "unsigned";
}
