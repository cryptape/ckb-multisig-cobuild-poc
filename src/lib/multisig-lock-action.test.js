import { buildScriptInfo, buildAction } from "./multisig-lock-action.js";
import { toJson } from "@ckb-cobuild/molecule";
import { decodeHex } from "@ckb-cobuild/hex-encoding";

test("buildScriptInfo", () => {
  const { schema, ...remaining } = buildScriptInfo();

  expect(schema).toMatch(
    `
array PubkeyHash [byte; 20];
vector PubkeyHashVec <PubkeyHash>;
table MultisigConfig {
    require_first_n: byte,
    threshold: byte,
    signer_pubkey_hashes: PubkeyHashVec,
}
array Signature [byte; 65];
struct PubkeyHashSignaturePair {
    pubkey_hash: PubkeyHash,
    signature: Signature,
}
vector PubkeyHashSignaturePairVec <PubkeyHashSignaturePair>;
table MultisigAction {
    config: MultisigConfig,
    signed: PubkeyHashSignaturePairVec,
}
    `.trim(),
  );
  expect(toJson(remaining)).toMatchSnapshot();
});

test("buildAction", () => {
  const action = buildAction(
    decodeHex("4cb239d9575deb2f62c998b74e7f0a8ee12daaaa"),
    {
      config: {
        require_first_n: 0,
        threshold: 2,
        signer_pubkey_hashes: [
          decodeHex("2910442b8955ddcc0d4c43919b2c98d7e5435434"),
          decodeHex("c8328aabcd9b9e8e64fbc566c4385c3bdeb219d7"),
          decodeHex("a656b9d66546084489ca979d95ca8d78ceb6764b"),
        ],
      },
      signed: [
        {
          pubkey_hash: decodeHex("2910442b8955ddcc0d4c43919b2c98d7e5435434"),
          signature: decodeHex(
            "743ea11867222e96da5b27dbdd63cec1bae3217e6a6c71a2a60abe602f43e7273f3cb25e88b6d00c19ea1dcb155dfd282a610e49d7a6e115aa8a058b451ca4a001",
          ),
        },
      ],
    },
  );
  expect(toJson(action)).toMatchSnapshot();
});
