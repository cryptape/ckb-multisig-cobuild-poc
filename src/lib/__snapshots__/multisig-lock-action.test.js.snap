// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildAction 1`] = `
Object {
  "data": "0xb70000000c0000005e000000520000001000000011000000120000000002030000002910442b8955ddcc0d4c43919b2c98d7e5435434c8328aabcd9b9e8e64fbc566c4385c3bdeb219d7a656b9d66546084489ca979d95ca8d78ceb6764b010000002910442b8955ddcc0d4c43919b2c98d7e5435434743ea11867222e96da5b27dbdd63cec1bae3217e6a6c71a2a60abe602f43e7273f3cb25e88b6d00c19ea1dcb155dfd282a610e49d7a6e115aa8a058b451ca4a001",
  "script_hash": "0x4ea8cdacceaba6dbfbd287535ed69395709246f1a9dd5485e30389bffed07b69",
  "script_info_hash": "0xa95d38f80309e7ac7a343f28d2f5f65a324ccd1fa5937cfc1615d1f9150eb2ac",
}
`;

exports[`buildScriptInfo 2`] = `
Object {
  "message_type": "MultisigAction",
  "name": "secp256k1_blake160_multisig_all",
  "script_hash": "0x1db3f661729719d90e8efe0cf33c6eba99225824536a6ce5a03ca0b3809c6344",
  "url": "https://github.com/nervosnetwork/ckb-system-scripts/blob/master/c/secp256k1_blake160_multisig_all.c",
}
`;
